system/jhtree/jhinplace.cpp

```diff
@@ -0,0 +1,2681 @@
+        case SqSingle:
+            out.appendf("si(%u) ", count);
+            break;
+        case SqSingleX:
+            out.appendf("sx(%u %02x) ", count, *cur);
+            cur++;
+            break;
+        }
+    }
+}
+
+
+void PartialMatch::gatherSingleTextCounts(StringBuffer & prev, unsigned * singleCounts)
+{
+    const byte * cur = squashedData.bytes();
+    const byte * end = cur + squashedData.length();
+    unsigned offset = rowOffset;
+    while (cur < end)
+    {
+        byte nextByte = *cur++;
+        SquashOp op = (SquashOp)(nextByte >> 5);
+        unsigned count = (nextByte & 0x1f);
+        if (count == 0)
+            count = 32 + *cur++;
+
+        switch (op)
+        {
+        case SqZero:
+        case SqSpace:
+            offset += count;
+            break;
+        case SqQuote:
+            if (count == 1)
+                singleCounts[*cur]++;
+#ifdef TRACE_BUILDING
+            //Horrible code to estimate how beneficial it would be to match strings that have previously occurred
+            //Should never be released enabled.
+            if (count >= 3 && next.ordinality() == 0)
+            {
+                StringBuffer s;
+                for (unsigned i=0; i < count; i++)
+                    s.appendf(" %2x", cur[i]);
+                if (strstr(prev, s.str()))
+                    DBGLOG("%p: @%u %u %s", this, offset, count, s.str());
+                prev.append("$").append(s.str());
+            }
+#endif
+            cur += count;
+            offset += count;
+            break;
+        case SqRepeat:
+            cur++;
+            offset += count;
+            break;
+        case SqSingle:
+            offset++;
+            break;
+        case SqSingleX:
+            cur++;
+            offset+=2;
+            break;
+        }
+    }
+
+    ForEachItemIn(i, next)
+        next.item(i).gatherSingleTextCounts(prev, singleCounts);
+}
+
+void PartialMatch::trace(unsigned indent)
+{
+    StringBuffer squashedText;
+    describeSquashed(squashedText);
+
+    StringBuffer clean;
+    StringBuffer dataHex;
+    for (unsigned i=0; i < data.length(); i++)
+    {
+        byte next = data.bytes()[i];
+        dataHex.appendhex(next, true);
+        if (isprint(next))
+            clean.append(next);
+        else
+            clean.append('.');
+    }
+
+    printf("%*s(%s[%s] %u:%u[%u] [%s]", indent, "", clean.str(), dataHex.str(), data.length(), squashedData.length(), getSize(), squashedText.str());
+    if (next.ordinality())
+    {
+        printf(", %u{\n", next.ordinality());
+        ForEachItemIn(i, next)
+            next.item(i).trace(indent+1);
+        printf("%*s}", indent, "");
+    }
+    printf(")\n");
+}
+
+//---------------------------------------------------------------------------------------------------------------------
+
+void PartialMatchBuilder::add(size32_t len, const void * data)
+{
+    //It is legal to add rows longer than keyLen, but cannot strip trailing nulls
```

Comment:
Why would you want to?

Reply:
Previous iteration of this code where it was also being used for compressing the payload - will remove the comment.